client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java

```diff
@@ -194,12 +204,29 @@ private void verifyIsAtCorrectOffset(WireCommands.SegmentRead segmentRead) {
     }
 
     /**
-     * Issues a request if there is enough room for another request, and we aren't already waiting on one
+     * Issues a request
+     *  - if there is enough room for another request, and we aren't already waiting on one and
+     *  - if we have not read up to the configured endOffset.
      */
     private void issueRequestIfNeeded() {
-        if (!receivedEndOfSegment && !receivedTruncated && buffer.capacityAvailable() >= readLength && outstandingRequest == null) {
-            outstandingRequest = asyncInput.read(offset + buffer.dataAvailable(), readLength);
+        //compute read length based on current offset upto which the events are read are read.
+        int updatedReadLength = computeReadLength(offset + buffer.dataAvailable(), readLength);
+        if (!receivedEndOfSegment && !receivedTruncated && updatedReadLength > 0 && buffer.capacityAvailable() >= updatedReadLength && outstandingRequest == null) {
+            outstandingRequest = asyncInput.read(offset + buffer.dataAvailable(), updatedReadLength);
+        }
+    }
+
+    /*
+     * Compute the read length based on the current fetch offset and the configured end offset.
+     */
+    private int computeReadLength(long currentFetchOffset, int currentReadLength) {
+        Preconditions.checkState(endOffset >= currentFetchOffset,
```

Comment:
This method will break the class if endOffset does not align to an event boundary. It will cause it to go into a tight loop fetching 0 bytes

Reply:
`issueRequestIfNeeded()` has an additional check which checks if the readLength > 0.